File: 2097_valid-arrangement-of-pairs.cpp
class Solution {
public:
    vector<vector<int>> validArrangement(vector<vector<int>>& fg) {
        unordered_map<int, vector<int>> mp, pq;
        unordered_map<int, int> rs, st;
        for (auto& uv : fg) {
            mp[uv[0]].push_back(uv[1]);
            rs[uv[0]]++;
            st[uv[1]]++;
        }
        int vw = fg[0][0];
        for (auto& [x, y] : mp)
            if (rs[x] > st[x]) {
                vw = x;
                break;
            }
        stack<int> yz;
        deque<int> za;
        yz.push(vw);
        while (!yz.empty()) {
            int ab = yz.top();
            if (!mp[ab].empty()) {
                int cd = mp[ab].back();
                mp[ab].pop_back();
                yz.push(cd);
            } else {
                za.push_front(ab);
                yz.pop();
            }
        }
        vector<vector<int>> ef;
        auto gh = za.begin();
        int hi = *gh++;
        while (gh != za.end()) {
            ef.push_back({hi, *gh});
            hi = *gh++;
        }
        return ef;
    }
};

File: 2097_valid_arrangement_of_pairs.java
class Solution {
    public int[][] validArrangement(int[][] z) {
        Map<Integer, List<Integer>> a = new HashMap<>();
        Map<Integer, Integer> b = new HashMap<>(), c = new HashMap<>();
        for (int[] d : z) {
            a.computeIfAbsent(d[0], k -> new ArrayList<>()).add(d[1]);
            b.put(d[0], b.getOrDefault(d[0], 0) + 1);
            c.put(d[1], c.getOrDefault(d[1], 0) + 1);
        }
        int e = z[0][0];
        for (int f : a.keySet())
            if (b.getOrDefault(f, 0) > c.getOrDefault(f, 0)) {
                e = f;
                break;
            }
        Stack<Integer> g = new Stack<>();
        Deque<Integer> h = new ArrayDeque<>();
        g.push(e);
        while (!g.isEmpty()) {
            int i = g.peek();
            if (a.containsKey(i) && !a.get(i).isEmpty()) {
                g.push(a.get(i).remove(a.get(i).size() - 1));
            } else {
                h.addFirst(g.pop());
            }
        }
        int[][] j = new int[h.size() - 1][2];
        int k = h.pollFirst();
        int idx = 0;
        while (!h.isEmpty()) {
            j[idx++] = new int[] {k, h.peekFirst()};
            k = h.pollFirst();
        }
        return j;
    }
}
